pkg/rc/replication_controller.go

```diff
@@ -633,81 +669,276 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+			OldNode: oldNode,
+			NewNode: newNode,
+		},
 	}
+
 	writeCtx, writeCancel := transaction.New(context.Background())
 	defer writeCancel()
-	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), modifyIndex, status)
+	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), 0, status)
 	if err != nil {
-		return util.Errorf("Could not write new node to store: %s", err)
+		return "", "", util.Errorf("Could not write new node to store: %s", err)
 	}
 
 	err = transaction.MustCommit(writeCtx, rc.txner)
 	if err != nil {
-		return util.Errorf("Could not commit CASTxn: %s", err)
+		return "", "", util.Errorf("Could not commit CASTxn: %s", err)
+	}
+
+	return newNode, oldNode, nil
+}
+
+func (rc *replicationController) scheduleWithSession(newNode types.NodeName) error {
+	rc.logger.NoFields().Infof("Scheduling with session on %s", newNode)
+	rc.mu.Lock()
+	manifest := rc.Manifest
+	rc.mu.Unlock()
+
+	manifestBytes, err := manifest.Marshal()
+	if err != nil {
+		return err
+	}
+
+	key, err := consul.PodPath(consul.INTENT_TREE, newNode, manifest.ID())
+	if err != nil {
+		return err
+	}
+
+	name := fmt.Sprintf("wait-on-%s-health-%s", newNode, rc.ID())
+
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	defer func() {
+		// We close the context on error to prevent a leak but leave it
+		// otherwise so it can be used to cancel the session. We write the
+		// defer this way so that reassignments to err are noticed and the
+		// final value is used.
+		if err != nil {
+			ctxCancel()
+		}
+	}()
+
+	// By passing ctx below, ctxCancel (defined above) can be used to destroy
+	// the session when the node transfer finishes or is stopped
+	_, session, err := consul.SessionContext(ctx, rc.consulClient, name)
+	if err != nil {
+		return err
+	}
+
+	writeCtx, writeCancel := transaction.New(context.Background())
+	defer writeCancel()
+
+	txnUnlocker, err := session.LockIfKeyNotExistsTxn(writeCtx, key, manifestBytes)
+	if err != nil {
+		return err
+	}
+
+	ok, resp, err := transaction.Commit(writeCtx, rc.txner)
+	switch {
+	case err != nil:
+		return err
+	case !ok:
+		// assign to err instead of just returning it so that deferred func sees
+		// the assignment
+		err = util.Errorf("transaction violation scheduling new node: %s", transaction.TxnErrorsToString(resp.Errors))
+		return err
 	}
 
+	rc.nodeTransfer.cancelSession = ctxCancel
+	rc.nodeTransfer.unlocker = txnUnlocker
+
 	return nil
 }
 
-func (rc *replicationController) transferPetNodes(ineligibleCurrent []types.NodeName) error {
-	errMsg := fmt.Sprintf("RC has scheduled %d ineligible nodes: %s", len(ineligibleCurrent), ineligibleCurrent)
-	err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+func (rc *replicationController) doBackgroundNodeTransfer() {
+	rc.isWatchingHealth = true
+	ok := rc.watchHealth()
+	if ok {
+		err := rc.finishTransfer()
+		if err != nil {
+			rc.logger.WithError(err).Errorln("could not do final node transfer transaction")
+		}
+	}
+
+	rc.isWatchingHealth = false
```

Comment:
I think these lines also are going to cause a race with the places where you do
```
if rc.IsWatchingHealth {
  close(rc.nodeTransfer.quit)
}
```
Probably there should be a mutex that guards the `isWatchingHealth` and `nodeTransfer` variables so that these operations are atomic. You might also be able to get away without using `isWatchingHealth` and just checking `nodeTransfer.quit != nil` or something like that.

Reply:
good call